File: scripts/scrape-bridging.ts
#!/usr/bin/env npx tsx
/**
 * Scrape BPS bridging API for all administrative levels.
 *
 * Output: scripts/bridging_provinsi.json, bridging_kabupaten.json,
 *         bridging_kecamatan.json, bridging_desa.json
 *
 * Usage: npx tsx scripts/scrape-bridging.ts
 */

import { readFileSync, writeFileSync, existsSync, statSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const BASE_URL = "https://sig.bps.go.id/rest-bridging/getwilayah";
const PERIODE = "2025s1";
const DELAY_MS = 50;
const MAX_RETRIES = 3;
const TIMEOUT_MS = 30_000;

/** 4 new Papua provinces not in BPS bridging */
const SKIP_PROVINCES = new Set(["92", "95", "96", "97"]);

interface BridgingRecord {
	kode_bps: string;
	kode_dagri: string;
	nama_bps: string;
}

interface ApiResponse {
	kode_bps: string;
	nama_bps: string;
	kode_dagri: string;
	nama_dagri: string;
}

function stripDots(code: string): string {
	return code.replaceAll(".", "");
}

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}

async function fetchWithTimeout(
	url: string,
	timeoutMs: number,
): Promise<Response> {
	const controller = new AbortController();
	const id = setTimeout(() => controller.abort(), timeoutMs);
	try {
		return await fetch(url, { signal: controller.signal });
	} finally {
		clearTimeout(id);
	}
}

async function fetchBridging(
	level: string,
	parent: string,
): Promise<ApiResponse[]> {
	const url = `${BASE_URL}?level=${level}&parent=${parent}&periode=${PERIODE}`;
	for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
		try {
			const resp = await fetchWithTimeout(url, TIMEOUT_MS);
			return (await resp.json()) as ApiResponse[];
		} catch (e) {
			console.error(
				`  RETRY ${attempt}/${MAX_RETRIES} for ${url}: ${e instanceof Error ? e.message : e}`,
			);
			if (attempt === MAX_RETRIES) {
				console.error(`  FAILED ${url}`);
				return [];
			}
			await sleep(2 ** attempt * 1000);
		}
	}
	return [];
}

function saveBridging(filename: string, data: BridgingRecord[]): void {
	const path = join(__dirname, filename);
	writeFileSync(path, JSON.stringify(data, null, undefined));
	console.log(`Saved ${data.length} records to ${path}`);
}

function loadIfExists(filename: string): BridgingRecord[] | null {
	const path = join(__dirname, filename);
	if (existsSync(path) && statSync(path).size > 10) {
		console.log(`Loading existing ${path}`);
		return JSON.parse(readFileSync(path, "utf-8")) as BridgingRecord[];
	}
	return null;
}

async function scrapeProvinsi(): Promise<BridgingRecord[]> {
	console.log("Fetching provinsi bridging...");
	const data = await fetchBridging("provinsi", "0");
	const result = data.map((item) => ({
		kode_bps: item.kode_bps,
		kode_dagri: stripDots(item.kode_dagri),
		nama_bps: item.nama_bps,
	}));
	console.log(`  Got ${result.length} provinsi`);
	return result;
}

async function scrapeKabupaten(
	provinsi: BridgingRecord[],
): Promise<BridgingRecord[]> {
	console.log("Fetching kabupaten bridging...");
	const result: BridgingRecord[] = [];
	for (let i = 0; i < provinsi.length; i++) {
		const prov = provinsi[i]!;
		if (SKIP_PROVINCES.has(prov.kode_bps)) continue;
		const data = await fetchBridging("kabupaten", prov.kode_bps);
		for (const item of data) {
			result.push({
				kode_bps: item.kode_bps,
				kode_dagri: stripDots(item.kode_dagri),
				nama_bps: item.nama_bps,
			});
		}
		await sleep(DELAY_MS);
		if ((i + 1) % 10 === 0)
			console.log(
				`  Progress: ${i + 1}/${provinsi.length} provinsi, ${result.length} kabupaten so far`,
			);
	}
	console.log(`  Got ${result.length} kabupaten`);
	return result;
}

async function scrapeKecamatan(
	kabupaten: BridgingRecord[],
): Promise<BridgingRecord[]> {
	console.log("Fetching kecamatan bridging...");
	const result: BridgingRecord[] = [];
	for (let i = 0; i < kabupaten.length; i++) {
		const kab = kabupaten[i]!;
		const data = await fetchBridging("kecamatan", kab.kode_bps);
		for (const item of data) {
			result.push({
				kode_bps: item.kode_bps,
				kode_dagri: stripDots(item.kode_dagri),
				nama_bps: item.nama_bps,
			});
		}
		await sleep(DELAY_MS);
		if ((i + 1) % 50 === 0)
			console.log(
				`  Progress: ${i + 1}/${kabupaten.length} kabupaten, ${result.length} kecamatan so far`,
			);
	}
	console.log(`  Got ${result.length} kecamatan`);
	return result;
}

async function scrapeDesa(
	kecamatan: BridgingRecord[],
): Promise<BridgingRecord[]> {
	console.log("Fetching desa bridging...");
	const result: BridgingRecord[] = [];
	const errors: string[] = [];
	for (let i = 0; i < kecamatan.length; i++) {
		const kec = kecamatan[i]!;
		const data = await fetchBridging("desa", kec.kode_bps);
		if (data.length === 0) errors.push(kec.kode_bps);
		for (const item of data) {
			result.push({
				kode_bps: item.kode_bps,
				kode_dagri: stripDots(item.kode_dagri),
				nama_bps: item.nama_bps,
			});
		}
		await sleep(DELAY_MS);
		if ((i + 1) % 200 === 0)
			console.log(
				`  Progress: ${i + 1}/${kecamatan.length} kecamatan, ${result.length} desa so far`,
			);
	}
	console.log(`  Got ${result.length} desa`);
	if (errors.length > 0)
		console.error(`  ERRORS for kecamatan: ${JSON.stringify(errors)}`);
	return result;
}

async function main(): Promise<void> {
	let provinsi =
		loadIfExists("bridging_provinsi.json") ?? (await scrapeProvinsi());
	if (!loadIfExists("bridging_provinsi.json"))
		saveBridging("bridging_provinsi.json", provinsi);

	let kabupaten =
		loadIfExists("bridging_kabupaten.json") ??
		(await scrapeKabupaten(provinsi));
	if (!loadIfExists("bridging_kabupaten.json"))
		saveBridging("bridging_kabupaten.json", kabupaten);

	let kecamatan =
		loadIfExists("bridging_kecamatan.json") ??
		(await scrapeKecamatan(kabupaten));
	if (!loadIfExists("bridging_kecamatan.json"))
		saveBridging("bridging_kecamatan.json", kecamatan);

	let desa =
		loadIfExists("bridging_desa.json") ?? (await scrapeDesa(kecamatan));
	if (!loadIfExists("bridging_desa.json"))
		saveBridging("bridging_desa.json", desa);

	console.log("\n=== Summary ===");
	console.log(`Provinsi:  ${provinsi.length}`);
	console.log(`Kabupaten: ${kabupaten.length}`);
	console.log(`Kecamatan: ${kecamatan.length}`);
	console.log(`Desa:      ${desa.length}`);
}

main().catch((err) => {
	console.error(err);
	process.exit(1);
});

File: scripts/scrape-bridging.py
#!/usr/bin/env python3
"""Scrape BPS bridging API for all administrative levels.

Output: scripts/bridging_provinsi.json, scripts/bridging_kabupaten.json,
        scripts/bridging_kecamatan.json, scripts/bridging_desa.json
"""
import json
import time
import urllib.request
import urllib.error
import sys
import os

BASE_URL = "https://sig.bps.go.id/rest-bridging/getwilayah"
PERIODE = "2025s1"
DELAY = 0.05  # 50ms between requests
MAX_RETRIES = 3
TIMEOUT = 30

# 4 new Papua provinces not in BPS bridging
SKIP_PROVINCES = {"92", "95", "96", "97"}

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))


def fetch(level: str, parent: str) -> list[dict]:
    """Fetch bridging data from BPS API with retries."""
    url = f"{BASE_URL}?level={level}&parent={parent}&periode={PERIODE}"
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            req = urllib.request.Request(url)
            resp = urllib.request.urlopen(req, timeout=TIMEOUT)
            return json.loads(resp.read())
        except (urllib.error.URLError, TimeoutError, OSError) as e:
            print(f"  RETRY {attempt}/{MAX_RETRIES} for {url}: {e}", file=sys.stderr)
            if attempt == MAX_RETRIES:
                print(f"  FAILED {url}", file=sys.stderr)
                return []
            time.sleep(2 ** attempt)
    return []


def strip_dots(code: str) -> str:
    """Remove dots from Kemendagri code: '11.09.07' -> '110907'"""
    return code.replace(".", "")


def scrape_provinsi() -> list[dict]:
    print("Fetching provinsi bridging...")
    data = fetch("provinsi", "0")
    result = []
    for item in data:
        result.append({
            "kode_bps": item["kode_bps"],
            "kode_dagri": strip_dots(item["kode_dagri"]),
            "nama_bps": item["nama_bps"],
        })
    print(f"  Got {len(result)} provinsi")
    return result


def scrape_kabupaten(provinsi: list[dict]) -> list[dict]:
    print("Fetching kabupaten bridging...")
    result = []
    for i, prov in enumerate(provinsi):
        if prov["kode_bps"] in SKIP_PROVINCES:
            continue
        data = fetch("kabupaten", prov["kode_bps"])
        for item in data:
            result.append({
                "kode_bps": item["kode_bps"],
                "kode_dagri": strip_dots(item["kode_dagri"]),
                "nama_bps": item["nama_bps"],
            })
        time.sleep(DELAY)
        if (i + 1) % 10 == 0:
            print(f"  Progress: {i + 1}/{len(provinsi)} provinsi, {len(result)} kabupaten so far")
    print(f"  Got {len(result)} kabupaten")
    return result


def scrape_kecamatan(kabupaten: list[dict]) -> list[dict]:
    print("Fetching kecamatan bridging...")
    result = []
    for i, kab in enumerate(kabupaten):
        data = fetch("kecamatan", kab["kode_bps"])
        for item in data:
            result.append({
                "kode_bps": item["kode_bps"],
                "kode_dagri": strip_dots(item["kode_dagri"]),
                "nama_bps": item["nama_bps"],
            })
        time.sleep(DELAY)
        if (i + 1) % 50 == 0:
            print(f"  Progress: {i + 1}/{len(kabupaten)} kabupaten, {len(result)} kecamatan so far")
    print(f"  Got {len(result)} kecamatan")
    return result


def scrape_desa(kecamatan: list[dict]) -> list[dict]:
    print("Fetching desa bridging...")
    result = []
    errors = []
    for i, kec in enumerate(kecamatan):
        data = fetch("desa", kec["kode_bps"])
        if not data:
            errors.append(kec["kode_bps"])
        for item in data:
            result.append({
                "kode_bps": item["kode_bps"],
                "kode_dagri": strip_dots(item["kode_dagri"]),
                "nama_bps": item["nama_bps"],
            })
        time.sleep(DELAY)
        if (i + 1) % 200 == 0:
            print(f"  Progress: {i + 1}/{len(kecamatan)} kecamatan, {len(result)} desa so far")
    print(f"  Got {len(result)} desa")
    if errors:
        print(f"  ERRORS for kecamatan: {errors}", file=sys.stderr)
    return result


def save(filename: str, data: list[dict]) -> None:
    path = os.path.join(SCRIPT_DIR, filename)
    with open(path, "w") as f:
        json.dump(data, f, ensure_ascii=False)
    print(f"Saved {len(data)} records to {path}")


def main():
    # Province level
    provinsi_path = os.path.join(SCRIPT_DIR, "bridging_provinsi.json")
    if os.path.exists(provinsi_path) and os.path.getsize(provinsi_path) > 10:
        print(f"Loading existing {provinsi_path}")
        with open(provinsi_path) as f:
            provinsi = json.load(f)
    else:
        provinsi = scrape_provinsi()
        save("bridging_provinsi.json", provinsi)

    # Kabupaten level
    kabupaten_path = os.path.join(SCRIPT_DIR, "bridging_kabupaten.json")
    if os.path.exists(kabupaten_path) and os.path.getsize(kabupaten_path) > 10:
        print(f"Loading existing {kabupaten_path}")
        with open(kabupaten_path) as f:
            kabupaten = json.load(f)
    else:
        kabupaten = scrape_kabupaten(provinsi)
        save("bridging_kabupaten.json", kabupaten)

    # Kecamatan level
    kecamatan_path = os.path.join(SCRIPT_DIR, "bridging_kecamatan.json")
    if os.path.exists(kecamatan_path) and os.path.getsize(kecamatan_path) > 10:
        print(f"Loading existing {kecamatan_path}")
        with open(kecamatan_path) as f:
            kecamatan = json.load(f)
    else:
        kecamatan = scrape_kecamatan(kabupaten)
        save("bridging_kecamatan.json", kecamatan)

    # Desa level
    desa_path = os.path.join(SCRIPT_DIR, "bridging_desa.json")
    if os.path.exists(desa_path) and os.path.getsize(desa_path) > 10:
        print(f"Loading existing {desa_path}")
        with open(desa_path) as f:
            desa = json.load(f)
    else:
        desa = scrape_desa(kecamatan)
        save("bridging_desa.json", desa)

    print("\n=== Summary ===")
    print(f"Provinsi:  {len(provinsi)}")
    print(f"Kabupaten: {len(kabupaten)}")
    print(f"Kecamatan: {len(kecamatan)}")
    print(f"Desa:      {len(desa)}")


if __name__ == "__main__":
    main()
